### ISSUE TITLE
New keyword `pin` for fixed sparse storage layouts

### ISSUE THREAD

**Issue by thedavidmeister** [ethereum/solidity#12513]
## Abstract

Apologies if this has already been requested, wasn't sure exactly what to search for.

Some kind of `pin` keyword for storage would make contracts using delegate call easier to work with.

Solidity sequentially orders storage slots based on the lexical ordering of things in in storage, including across contracts.

This means that working on projects where the storage layout changes over time requires some awkwardness like https://docs.openzeppelin.com/contracts/4.x/api/utils#StorageSlot and https://github.com/OpenZeppelin/openzeppelin-contracts-upgradeable/blob/master/contracts/token/ERC20/ERC20Upgradeable.sol#L362

The `StorageSlot` from OZ takes a hardcoded hash like:

```
    // This is the keccak-256 hash of "eip1967.proxy.rollback" subtracted by 1
    bytes32 private constant _ROLLBACK_SLOT = 0x4910fdfa16fed3260ed0e7147f7cc6da11a60208b5b9406d12a635614ffd9143;
```

To put primitives like `address`, `bool`, `bytes32` and `uint256` into a known _non-sequential_ slot.

Solidity itself could do the same thing but better as it could support mapping and everything else, which would then also remove the need for "gaps" where the contract forces sequential spacing between storage locations by putting say 50 unused slots in every contract "just in case".

Current:

```
contract Foo {
  uint256 private foo;
  mapping (address => uint256) private bar;
}
```

Pinned:

```
contract Foo {
    uint256 private foo pin "eip1967.proxy.rollback"; // foo is at the rollback slot as above rather than 0
    mapping (address => uint256) private bar pin "something.else"; // normal mapping logic with hashes of keys but sans sequential offset
}
```

Bonus if having multiple things with the same pin would work the same as current sequential logic, i.e. a sequential offset from the start of the pin location, to support packing etc. as per normal solidity

Pinned packed:

```
contract Foo {
   uint16 private foo pin "some.pin";
   uint16 private bar pin "some.pin"; // bar packs with foo as they share the same pin, doesn't overwrite
}
```

Not having the `pin` keyword falls back to slot `0` which behaves exactly as solidity does now. Also integer literals for the pin are supported so e.g. hex hashes work fine too.

```
contract Foo {
  uint256 private foo pin 0;
  uint256 private bar; // shares the pin with foo as `0` is the fallback
}
```

## Motivation

So one way to interject new storage across versions (e.g. think of inheritance graphs where append-only is not possible) would be to  have the same pin for everything that was first deployed at the same time.

First deploy:

```
contract V1 {
  uint foo pin "v1";
  uint bar pin "v1";
}
```

Second deploy:

```
contract V2 {
  uint foo pin "v1";
  uint baz pin "v2";
  uint bar pin "v1";
}
```

Without pinning, the addition of `baz` pushes `bar` down, but with pinning `bar` remains in the same place, in the second slot after `foo` at their shared pin.

Alternatively, the pins could be named uniquely but the same within a base contract across all versions, so adding additional storage to the base contract would not impact any other contract that inherits from it. This is basically the "gap" strategy but with the sparseness (never run out of slots) of hashing and automatically managed by solidity rather than a hand-hacked 50 item array.

First deploy:

```
contract Av1 {
  uint foo pin "A";
}

contract Bv1 {
  uint foo pin "B":
}

contract C is A, B { }
```

Second deploy:

```
contract Av2 {
  uint foo pin "A";
  uint bar pin "A";
}

contract Bv1 {
  uint foo pin "B";
}

contract C is A, B { }
```

There's probably other good conventions and strategies that make pinning useful, but this is the basic idea.

## Specification

New keyword `pin` for any/all storage.

`pin` must be followed by either an integer literal or string, the latter is hashed to normalize it to an integer.

If `pin` is absent, fallback to `0`.

The pin integer defines a new space/offset within/from which sequential storage behaves exactly as it does now re: packing, hashing mapping keys, etc. across inheritance, it's all the same as it is now but from the `pin` defined offset.

All offsets are computed at compile time so as not to introduce additional gas costs.

## Backwards Compatibility

If `pin` is absent fallback to `0`.

**Comment by ekpyron** [comment]
Oh yes, this, resp. similar requests have come up before, but there is no consensus on any of it yet, so thank you very much for the suggestion!
Related older issues are (maybe among others): https://github.com/ethereum/solidity/issues/4519 https://github.com/ethereum/solidity/issues/597 https://github.com/ethereum/solidity/issues/8353 https://github.com/ethereum/solidity/issues/8490

**Comment by thedavidmeister** [comment]
yes it seems to me like the upgrade problems are introduced by design choices in solidity, not fundamental to the evm, so i'm not surprised it has come up before :)

**Comment by thedavidmeister** [comment]
so the obvious next question is, how to reach consensus @ekpyron because these conversations date back 6 years, how are they getting stalled?

is this a "PR welcome" situation, or is there something blocking code?

**Comment by ekpyron** [comment]
The main issue is to reach a decision about whether and how this should be allowed, to what extend and with what concrete syntax - once a decision was reached about all that, it would probably be rather trivial and quick to implement this. But unless there is agreement about those design questions, it won't happen.
So at this stage a PR would not help as much as further commenting on the relevant issues and trying to get people to agree on one approach.

It got stalled in the past because there's quite different opinions on how this should be done, resp. whether it should be done at all or whether mechanisms like using inline assembly to synthesize storage pointers are to be considered sufficient.

We can try to prioritize this again and try to get to a decision - but I'm sure @chriseth will want to weigh in on this and he's on vacation until February.

**Comment by thedavidmeister** [comment]
@ekpyron of course there are workarounds like assembly, also the problem is introduced by a design decision in solidity so it feels like something solidity should handle

**Comment by thedavidmeister** [comment]
@ekpyron  so maybe the steps are like:

1. agree that contract upgradeability issues described above are due to a leaky abstraction in solidity to be fixed rather than worked around with hand rolled "gaps" or inline assembly
2. agree that an acceptable solution exists somewhere in the last 6 years of discourse
3. agree on some criteria that allow us to definitively exclude some proposed options
4. some kind of decision making process re: the final shortlist
5. implement the solution

**Comment by ekpyron** [comment]
Something like that, yes.

**Comment by wranai** [comment]
I came here to request the same thing. My idea was to add a more generic `@ slot`  modifier at the end of the declaration, something like:

```solidity
contract Foo {
    uint256 x; // slot 0 as usual
    uint256 private foo = 0x123 @ keccak256("eip1967.proxy.rollback") - 1; // foo is at the rollback slot, exempt from the sequential numbering
    mapping (address => uint256) private bar; // continued with slot 1
}
```

The compiler should pay attention to skip the numbers that are specified, or maybe `slot` could be restricted to numbers larger than e.g. 2**32 to avoid clashes.

EDIT: And now I just saw it's **exactly** what was suggested here: https://github.com/ethereum/solidity/issues/597#221609070; I guess it's an intuitive way to go about it multiple people thought of the same thing.
